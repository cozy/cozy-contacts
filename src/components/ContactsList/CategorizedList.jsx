import React from 'react'
import PropTypes from 'prop-types'

import { useI18n } from 'cozy-ui/transpiled/react/I18n'

import ContactHeaderRow from './ContactHeaderRow'
import ContactsSubList from './ContactsSubList'
import { categorizeContacts } from '../../helpers/contactList'

const CategorizedList = ({ contacts }) => {
  const { t } = useI18n()
  const categorizedContacts = categorizeContacts(contacts, t('empty-list'))

  return (
    <ol className="list-contact">
      {Object.entries(categorizedContacts).map(([header, contacts]) => (
        <li key={`cat-${header}`}>
          <ContactHeaderRow key={header} header={header} />
          <ContactsSubList contacts={contacts} />
        </li>
      ))}
    </ol>
  )
}

CategorizedList.propTypes = {
  contacts: PropTypes.array.isRequired
}

export default CategorizedList
