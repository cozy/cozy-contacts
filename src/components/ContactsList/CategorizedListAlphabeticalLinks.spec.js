import React from 'react'

import renderer from 'react-test-renderer'
import { mount } from 'enzyme'

import AppLike from '../../tests/Applike'
import CategorizedList from './CategorizedList'

const defaultSelectionProps = {
  areAllContactsSelected: false,
  handleAllContactSelection: () => {}
}

const aContact = {
  name: { familyName: 'Cozy', givenName: 'Cloud' },
  indexes: { byFamilyNameGivenNameEmailCozyUrl: 'cozycloud' },
  _id: 0
}

const anonymousContact = {
  name: { familyName: '', givenName: '' },
  indexes: { byFamilyNameGivenNameEmailCozyUrl: {} },
  _id: 1
}

const contactList = [
  {
    name: { familyName: 'Co', givenName: 'Zy' },
    indexes: { byFamilyNameGivenNameEmailCozyUrl: 'cozy' },
    _id: 2
  },
  {
    name: { familyName: 'Vic', givenName: 'Tor' },
    indexes: { byFamilyNameGivenNameEmailCozyUrl: 'victor' },
    _id: 3
  },
  {
    name: { familyName: 'Gre', givenName: 'Noble' },
    indexes: { byFamilyNameGivenNameEmailCozyUrl: 'grenoble' },
    _id: 4
  }
]

const mountACategorizedList = contacts => {
  const categorizedListInstance = (
    <AppLike>
      <CategorizedList contacts={contacts} {...defaultSelectionProps} />
    </AppLike>
  )
  return mount(categorizedListInstance)
}

const checkContactListMatchExpectedLinks = (wrapper, expectedLetters) => {
  expect(wrapper).toBeDefined()
  expect(wrapper.length).toEqual(expectedLetters.length)
  for (let i = 0; i < expectedLetters.length; i++) {
    expect(wrapper.at(i).text()).toBe(expectedLetters[i])
  }
}

describe('CategorizedContactList alphabetical links', () => {
  it('should represent a contact initial letter', () => {
    const categorizedList = mountACategorizedList([aContact])
    const alphabeticalLinks = categorizedList.find('Button')
    checkContactListMatchExpectedLinks(alphabeticalLinks, ['C'])
  })

  it('should represent the exact list of initials of a list of contacts', () => {
    const categorizedList = mountACategorizedList(contactList)
    const alphabeticalLinks = categorizedList.find('Button')
    checkContactListMatchExpectedLinks(alphabeticalLinks, ['C', 'G', 'V'])
  })

  it('should provide an ordered list of alphabetical links', () => {
    const categorizedList = mountACategorizedList([
      contactList[2],
      contactList[0],
      contactList[1]
    ])
    const alphabeticalLinks = categorizedList.find('Button')
    checkContactListMatchExpectedLinks(alphabeticalLinks, ['C', 'G', 'V'])
  })

  it('should provide only one alphabetical link for every contact having the same initial', () => {
    const categorizedList = mountACategorizedList(contactList.concat(aContact))
    const alphabeticalLinks = categorizedList.find('Button')
    checkContactListMatchExpectedLinks(alphabeticalLinks, ['C', 'G', 'V'])
  })

  it('should not provide alphabetical link for anonymous contacts', () => {
    const categorizedList = mountACategorizedList(
      contactList.concat(anonymousContact)
    )
    const alphabeticalLinks = categorizedList.find('Button')
    checkContactListMatchExpectedLinks(alphabeticalLinks, ['C', 'G', 'V'])
  })

  it('should match the categorized list snapshot', () => {
    const tree = renderer
      .create(
        <AppLike>
          <CategorizedList contacts={contactList} {...defaultSelectionProps} />
        </AppLike>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
