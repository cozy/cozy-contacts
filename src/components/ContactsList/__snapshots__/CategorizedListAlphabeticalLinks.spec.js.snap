// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategorizedContactList alphabetical links should match the categorized list snapshot 1`] = `
Array [
  <div>
    <button
      className="styles__c-btn___-2Vnj styles__c-btn--subtle___OknKf styles__c-btn--center___16_Xh alphabetical-link"
      onClick={[Function]}
      title={null}
      type="submit"
    >
      <span>
        <span
          className={null}
        >
          C
        </span>
      </span>
    </button>
    <button
      className="styles__c-btn___-2Vnj styles__c-btn--subtle___OknKf styles__c-btn--center___16_Xh alphabetical-link"
      onClick={[Function]}
      title={null}
      type="submit"
    >
      <span>
        <span
          className={null}
        >
          G
        </span>
      </span>
    </button>
    <button
      className="styles__c-btn___-2Vnj styles__c-btn--subtle___OknKf styles__c-btn--center___16_Xh alphabetical-link"
      onClick={[Function]}
      title={null}
      type="submit"
    >
      <span>
        <span
          className={null}
        >
          V
        </span>
      </span>
    </button>
  </div>,
  <div
    className="styles__Table___x3ZsI"
  >
    <ul
      className="MuiList-root MuiList-padding"
    >
      <li
        className="MuiListSubheader-root MuiListSubheader-sticky MuiListSubheader-gutters"
      >
        c
      </li>
      <ul
        className="MuiList-root MuiList-padding"
      >
        <li
          className="MuiListItem-root u-c-pointer MuiListItem-gutters"
          data-testid="contact-listItem"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="contact-selection"
            onClick={[Function]}
          >
            <span
              data-input="checkbox"
            >
              <input
                checked={false}
                readOnly={true}
                type="checkbox"
              />
              <label />
            </span>
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-identity u-flex u-flex-items-center u-ellipsis"
          >
            <div
              className="styles__c-avatar___PpDI- styles__c-avatar--text___2dvna"
              style={
                Object {
                  "--circleSize": "32px",
                  "backgroundColor": "var(--lavender)",
                }
              }
            >
              
              <span
                className="styles__c-avatar-initials___310qC"
              >
                ZC
              </span>
            </div>
            <p
              className="MuiTypography-root u-ml-1 MuiTypography-body1 MuiTypography-noWrap MuiTypography-gutterBottom"
              inline={true}
            >
              <span
                className=""
              >
                Zy
                 
              </span>
              <span
                className="u-fw-bold"
              >
                Co
                 
              </span>
            </p>
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-email u-ellipsis"
          >
            —
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-phone u-ellipsis"
          >
            —
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-cozyurl u-ellipsis"
          >
            —
          </div>
        </li>
      </ul>
    </ul>
    <ul
      className="MuiList-root MuiList-padding"
    >
      <li
        className="MuiListSubheader-root MuiListSubheader-sticky MuiListSubheader-gutters"
      >
        v
      </li>
      <ul
        className="MuiList-root MuiList-padding"
      >
        <li
          className="MuiListItem-root u-c-pointer MuiListItem-gutters"
          data-testid="contact-listItem"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="contact-selection"
            onClick={[Function]}
          >
            <span
              data-input="checkbox"
            >
              <input
                checked={false}
                readOnly={true}
                type="checkbox"
              />
              <label />
            </span>
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-identity u-flex u-flex-items-center u-ellipsis"
          >
            <div
              className="styles__c-avatar___PpDI- styles__c-avatar--text___2dvna"
              style={
                Object {
                  "--circleSize": "32px",
                  "backgroundColor": "var(--puertoRico)",
                }
              }
            >
              
              <span
                className="styles__c-avatar-initials___310qC"
              >
                TV
              </span>
            </div>
            <p
              className="MuiTypography-root u-ml-1 MuiTypography-body1 MuiTypography-noWrap MuiTypography-gutterBottom"
              inline={true}
            >
              <span
                className=""
              >
                Tor
                 
              </span>
              <span
                className="u-fw-bold"
              >
                Vic
                 
              </span>
            </p>
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-email u-ellipsis"
          >
            —
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-phone u-ellipsis"
          >
            —
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-cozyurl u-ellipsis"
          >
            —
          </div>
        </li>
      </ul>
    </ul>
    <ul
      className="MuiList-root MuiList-padding"
    >
      <li
        className="MuiListSubheader-root MuiListSubheader-sticky MuiListSubheader-gutters"
      >
        g
      </li>
      <ul
        className="MuiList-root MuiList-padding"
      >
        <li
          className="MuiListItem-root u-c-pointer MuiListItem-gutters"
          data-testid="contact-listItem"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="contact-selection"
            onClick={[Function]}
          >
            <span
              data-input="checkbox"
            >
              <input
                checked={false}
                readOnly={true}
                type="checkbox"
              />
              <label />
            </span>
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-identity u-flex u-flex-items-center u-ellipsis"
          >
            <div
              className="styles__c-avatar___PpDI- styles__c-avatar--text___2dvna"
              style={
                Object {
                  "--circleSize": "32px",
                  "backgroundColor": "var(--pumpkinOrange)",
                }
              }
            >
              
              <span
                className="styles__c-avatar-initials___310qC"
              >
                NG
              </span>
            </div>
            <p
              className="MuiTypography-root u-ml-1 MuiTypography-body1 MuiTypography-noWrap MuiTypography-gutterBottom"
              inline={true}
            >
              <span
                className=""
              >
                Noble
                 
              </span>
              <span
                className="u-fw-bold"
              >
                Gre
                 
              </span>
            </p>
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-email u-ellipsis"
          >
            —
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-phone u-ellipsis"
          >
            —
          </div>
          <div
            className="styles__TableCell___yJCq7 contact-cozyurl u-ellipsis"
          >
            —
          </div>
        </li>
      </ul>
    </ul>
  </div>,
]
`;
